src/runtime_src/core/pcie/driver/linux/xocl/subdev/ert_user.c

```diff
@@ -0,0 +1,986 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include "../xocl_drv.h"
+
+#define	ERT_MAX_SLOTS 	128
+
+#define	ERT_STATE_GOOD	0x1
+#define	ERT_STATE_BAD	0x2
+
+
+//#define	SCHED_VERBOSE	1
+
+#ifdef SCHED_VERBOSE
+#define	ERTUSER_ERR(ert_user, fmt, arg...)	\
+	xocl_err(ert_user->dev, fmt "", ##arg)
+#define	ERTUSER_INFO(ert_user, fmt, arg...)	\
+	xocl_info(ert_user->dev, fmt "", ##arg)	
+#define	ERTUSER_DBG(ert_user, fmt, arg...)	\
+	xocl_info(ert_user->dev, fmt "", ##arg)
+#else
+#define	ERTUSER_ERR(ert_user, fmt, arg...)	\
+	xocl_err(ert_user->dev, fmt "", ##arg)
+#define	ERTUSER_INFO(ert_user, fmt, arg...)	\
+	xocl_info(ert_user->dev, fmt "", ##arg)	
+#define	ERTUSER_DBG(ert_user, fmt, arg...)
+#endif
+
+
+#define sched_debug_packet(packet, size)				\
+({									\
+	int i;								\
+	u32 *data = (u32 *)packet;					\
+	for (i = 0; i < size; ++i)					    \
+		DRM_INFO("packet(0x%p) execbuf[%d] = 0x%x\n", data, i, data[i]); \
+})
+
+struct ert_user_event {
+	struct mutex		  lock;
+	void			 *client;
+	int			  state;
+};
+
+struct ert_user_command {
+	struct kds_command *xcmd;
+	struct list_head    list;
+	uint32_t	slot_idx;
+};
+
+struct xocl_ert_user {
+	struct device		*dev;
+	struct platform_device  *pdev;
+	void __iomem		*cq_base;
+	uint64_t		cq_range;
+	void __iomem		*csr_base;
+	uint64_t		csr_range;
+	u32			intr_base;
+	u32			intr_num;
+	bool			polling_mode;
+	struct mutex 		lock;
+	struct kds_ert		ert;
+
+
+	/* Configure dynamically */ 
+	unsigned int		num_slots;
+	bool			cq_intr;
+	bool			config;
+	bool			ctrl_busy;
+	// Bitmap tracks busy(1)/free(0) slots in command_queue
+	DECLARE_BITMAP(slot_status, ERT_MAX_SLOTS);
+	struct xocl_ert_sched_privdata ert_cfg_priv;
+
+	struct ert_user_command 	  *submit_queue[ERT_MAX_SLOTS];
+	spinlock_t		  sq_lock;
+	u32			  num_sq;
+
+	struct list_head	  pq;
+	spinlock_t		  pq_lock;
+	u32			  num_pq;
+	/*
+	 * Pending Q is used in thread that is submitting CU cmds.
+	 * Other Qs are used in thread that is completing them.
+	 * In order to prevent false sharing, they need to be in different
+	 * cache lines. Hence we add a "padding" in between (assuming 128-byte
+	 * is big enough for most CPU architectures).
+	 */
+	u64			  padding[16];
+	/* run queue */
+	struct list_head	  rq;
+	u32			  num_rq;
+	/* completed queue */
+	struct list_head	  cq;
+	u32			  num_cq;
+	struct semaphore	  sem;
+
+	u32			  stop;
+	bool			  bad_state;
+
+	struct ert_user_event	  ev;
+
+	struct task_struct	  *thread;
+};
+
+static const unsigned int no_index = -1;
+static void ert_user_reset(struct xocl_ert_user *ert_user);
+
+static ssize_t name_show(struct device *dev,
+			   struct device_attribute *attr, char *buf)
+{
+	//struct xocl_ert_user *ert_user = platform_get_drvdata(to_platform_device(dev));
+	return sprintf(buf, "ert_user");
+}
+
+static DEVICE_ATTR_RO(name);
+
```

Comment:
This is not a change request.

Maybe we could a sysfs node to show current configuration for debugging purpose.